File: app/src/main/kotlin/metamark/parser/GrammarTerm.kt
package metamark.parser

import metamark.parser.antlr.MetamarkLexer
import metamark.parser.antlr.MetamarkParser
import metamark.parser.antlr.MetamarkParser.Grammar_Context
import org.antlr.v4.runtime.CharStreams
import org.antlr.v4.runtime.CommonTokenStream

data class GrammarTerm(
    val definitions: List<DefinitionTerm>,
) {
    companion object {
        fun parse(
            source: String,
        ): GrammarTerm {
            val sourceName = "__grammar__"

            val lexer = MetamarkLexer(CharStreams.fromString(source, sourceName))
            val tokenStream = CommonTokenStream(lexer)
            val parser = MetamarkParser(tokenStream)

            return GrammarTerm.build(parser.grammar_())
        }

        private fun build(ctx: Grammar_Context): GrammarTerm = GrammarTerm(
            definitions = ctx.definition().map {
                DefinitionTerm.build(it)
            },
        )
    }
}

File: src/main/GrammarTerm.ts
import {parseAsContext} from "./utils";
import {Grammar_Context} from "../generated_parser/MetamarkParser";
import {DefinitionTerm} from "./DefinitionTerm";

export class GrammarTerm {
    static parse(source: string) {
        return GrammarTerm.build(
            parseAsContext(
                (parser) => parser.grammar_(),
                source,
            ),
        );
    }

    static build(ctx: Grammar_Context): GrammarTerm {
        return new GrammarTerm(ctx.definition().map(
            (it) => DefinitionTerm.build(it))
        );
    }

    readonly definitions: ReadonlyArray<DefinitionTerm>;

    constructor(definitions: ReadonlyArray<DefinitionTerm>) {
        this.definitions = definitions;
    }
}
